### PR TITLE
fix #6282: calculate correct count for PoolArena's tiny/small/normal allocation

### LINKED ISSUE #6282: Incorrect allocations value for PoolArena (tiny / small / normal)
I found some PoolArena allocations value was incorrect, more specifically is:

    long numTinyAllocations();
    long numSmallAllocations();
    long numNormalAllocations();

and

    long numNormalDeallocations();

When I alloc Pooled ByteBuf & and release all these Bufs,

### Expected behavior
 
PoolArena's Metric SHOULD meet the following conditions:
```
           numTinyAllocations() == numTinyDeallocations()
     &&   numSmallAllocations() == numSmallDeallocations()
     &&   numNormalAllocations() == numNormalDeallocations()
```
### Actual behavior

BUT now result:
```
           numTinyAllocations() < numTinyDeallocations()
     &&   numSmallAllocations() < numSmallDeallocations()
     &&   numNormalAllocations() > numNormalDeallocations()
```
It seems some tiny & small allocations increase normal's counter,  I export PoolArenaMetric as MBean by [code](https://github.com/isdom/jocean-http/blob/9d8d528a2a3ee6bf2ca12f04054523dec94d3507/jocean-http/src/main/java/org/jocean/http/util/NettyStats.java) and show MBean by Web using zkoss, see below:

In this case, 

`numAllocations(2404) == numDeallocations(2404)`

BUT  
```
   (numTinyDeallocations - numTinyAllocations)       //  == 17
+ (numSmallDeallocations - numSmallAllocations)       //  == 9
```
equals

` (numNormalAllocations - numNormalDeallocations)       // == 26`

### Steps to reproduce

I start test code with VM flag (disable thread local cache): 
```
-XX:MaxDirectMemorySize=96M
-Dio.netty.allocator.tinyCacheSize=0 
-Dio.netty.allocator.smallCacheSize=0 
-Dio.netty.allocator.normalCacheSize=0 
-Dio.netty.allocator.type=pooled
```
then alloc some ByteBuf and release Bufs.

### Minimal yet complete reproducer code (or URL to code)

TestCase to reproduce: https://github.com/isdom/jocean-http/blob/6bc6cfa9ae1e71395a4dd52355b1b0a8365bb530/jocean-http/src/test/java/org/jocean/netty/buffer/PooledByteBufAllocatorTestCase.java

and make sure run this test with VM flag: -XX:MaxDirectMemorySize=96M

It fail at:

```
assertEquals(metric.numTinyDeallocations(), metric.numTinyAllocations());
assertEquals(metric.numSmallDeallocations(), metric.numSmallAllocations());
assertEquals(metric.numNormalDeallocations(), metric.numNormalAllocations());
```
output : 
```
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
```

If this is confirmed as a issue，I can open a PR to fix it.

### Netty version

netty-all-4.1.7.Final

### JVM version (e.g. `java -version`)

java version "1.8.0_102"
Java(TM) SE Runtime Environment (build 1.8.0_102-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)

### OS version (e.g. `uname -a`)

Linux xxx 3.10.0-327.22.2.el7.x86_64 #1 SMP Thu Jun 23 17:05:11 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by isdom**
Motivation:

Disable ThreadLocal Cache, then allocate Pooled ByteBuf and release all these buffers, PoolArena's  tiny/small/normal allocation count incorrect.

Modifications:

- Calculate PoolArena's tiny/small/normal allocation one time 
- Add testAllocationCounter TestCase

Result:

Fixes #6282 .

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 1]
```diff
@@ -190,35 +190,34 @@ private void allocate(PoolThreadCache cache, PooledByteBuf<T> buf, final int req
...
+                        assert handle >= 0;
+                        s.chunk.initBufWithSubpage(buf, handle, reqCapacity);
+                        return;
                     }
-                    return;
                 }
+                allocateNormal(buf, reqCapacity, normCapacity);
+                return;
+            } finally {
+                incTinySmallNormalAllocation(normCapacity);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @isdom I would prefer if we not need to acquire the `synchronized(this)` two times when we increment for normal allocations. Can you please refactor to ensure we not do ?

Comment (isdom) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: OK

Comment (isdom) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @normanmaurer Could you give me some advice for my commit: [acquire the synchronized(this) once when we increment for normal allocations](https://github.com/netty/netty/pull/6288/commits/e0b79fe7e7e70ccaf5c3178be773c0a8a552a1fe)?

**Comment by normanmaurer** [comment]
@isdom thanks a lot... I squashed your commits and made a tiny change to minimize synchronized scope.

Cherry-picked into 4.1 (f10f8a31318a2e408b979de6a8ed49caa615d86a) and 4.0 (025e656fd954d8aa93a7b025ef8d30235c5f0eca)

**Comment by isdom** [comment]
@normanmaurer thanks for your adjustment. the code looks much better.

**Comment by normanmaurer** [comment]
@isdom no worries... thanks again for the help and the report!

**Pull request closed without merging**